src/index.js

```diff
@@ -1,174 +1,231 @@
    * @api public
    */
 
-  function empty(room, fn = noop) {
-    if (room && rooms[room]) {
-      if (!isArray(room)) clear(room)
-      else for (const rm of room) clear(rm)
+  function empty(room) {
+    if (isArray(room)) {
+      let promises = [];
+      for (let rm in rooms) promises.push(clear(rm));
+      return Promise.all(promises);
     }
-    setImmediate(fn, null, null)
+    return clear(room);
   }
 
   /**
    * Reset all rooms or a particular room.
-   *
-   * @param {String} room
-   * @param {Function} [fn]
+   * Helper function for empty
+   * @param {String} room - the room to clear
    * @api public
    */
-
-  function clear(room, fn = noop) {
-    if ('function' === typeof room) {
-      rooms = {}, sids = {}
-    }
-    for (const id in rooms[room]) {
-      delete sids[id][room]
+  function clear(room) {
+    try {
+      if (!room) {
+        rooms.clear();
+        sids.clear();
+        return Promise.resolve({success: true});
+      }
+      let sockets = rooms.get(room);
+      if (sockets instanceof Set) {
+        for (const id of sockets) {
+          let socketRooms = sids.get(id);
+          socketRooms.delete(room);
+          if (socketRooms.size === 0) {
+            sids.delete(id);
+          }
+          rooms.delete(room);
+          // TODO update wild 
+          // wild.remove(room)
+        }
+      }
+      return Promise.resolve({success: true});
+    } catch (e) {
+      return Promise.reject(e);
     }
-    delete rooms[room]
-    wild.remove(room)
-    setImmediate(fn, null, null)
   }
 
   /**
    * Broadcast a packet.
-   *
-   * Options:
-   *  - `except` {Array} sids that should be excluded
-   *  - `rooms` {Array} list of rooms to broadcast to
-   *  - `method` {String} 'write' or 'send' if primus-emitter is present
-   *
-   * @param {Object} data
-   * @param {Object} opts
-   * @param {Object} clients Connected clients
+   * Used for broadcast
+   * @param {*} data - the data to transmit
+   * @param {Object} options - options to use
+   * @param {String[]} options.except - spark IDs to exclude
+   * @param {String[]} options.rooms - list of the rooms to broadcast to
+   * @param {String} options.method - 'write' or 'send' if primus-emitter is present
+   * @param {Function} options.transformer - the transformer function to call on the data
+   * @param {Object} clients -  Connected clients
    * @api public
    */
 
   function broadcast(data, options = {}, clients) {
     let rms = options.rooms || []
     if (!rms.length) {
-      return send(sids, clients, data, options)
-    }
-    for (const room of rms) {
-      let ids = rooms[room]
-      if (ids) send(ids, clients, data, options)
-      wild.match(room, key =>
-        send(rooms[key], clients, data, options)
+      let ids = new Set();
+      sids.forEach((value, key) => {
+        ids.add(key);
+      });
+      return send(ids, clients, data, options)
     }
+    let promises = [];
+    rooms.forEach((value, room) => {
```

Comment:
This doesn't look correct to me. We should only loop trough the rooms that have been passed via `options.rooms`.